src/arduino.cc/builder/phases/sizer.go

```diff
@@ -0,0 +1,160 @@
+	buildProperties := ctx.BuildProperties
+	verbose := ctx.Verbose
+	warningsLevel := ctx.WarningsLevel
+	logger := ctx.GetLogger()
+
+	err := checkSize(buildProperties, verbose, warningsLevel, logger)
+	if err != nil {
+		return i18n.WrapError(err)
+	}
+
+	return nil
+}
+
+func checkSize(buildProperties properties.Map, verbose bool, warningsLevel string, logger i18n.Logger) error {
+
+	properties := buildProperties.Clone()
+	properties[constants.BUILD_PROPERTIES_COMPILER_WARNING_FLAGS] = properties[constants.BUILD_PROPERTIES_COMPILER_WARNING_FLAGS+"."+warningsLevel]
+
+	maxTextSizeString := properties["upload.maximum_size"]
+	maxDataSizeString := properties["upload.maximum_data_size"]
+
+	if maxTextSizeString == "" {
+		return nil
+	}
+
+	logLevel := constants.LOG_LEVEL_INFO
+
+	maxTextSize, _ := strconv.Atoi(maxTextSizeString)
+	maxDataSize := -1
+
+	if maxDataSizeString != "" {
+		maxDataSize, _ = strconv.Atoi(maxDataSizeString)
+	}
+
+	var textSize, dataSize, eepromSize int
+
+	err := execSizeReceipe(properties, logger, &textSize, &dataSize, &eepromSize)
+	if err != nil {
+		return err
+	}
+
+	logger.Fprintln(os.Stdout, logLevel, "Sketch uses {0} bytes ({2}%%) of program storage space. Maximum is {1} bytes.", strconv.Itoa(textSize), strconv.Itoa(maxTextSize), strconv.Itoa(textSize*100/maxTextSize))
+	if dataSize > 0 {
+		if maxDataSize > 0 {
+			logger.Fprintln(os.Stdout, logLevel, "Global variables use {0} bytes ({2}%%) of dynamic memory, leaving {3} bytes for local variables. Maximum is {1} bytes.", strconv.Itoa(dataSize), strconv.Itoa(maxDataSize), strconv.Itoa(dataSize*100/maxDataSize), strconv.Itoa(maxDataSize-dataSize))
+		} else {
+			logger.Fprintln(os.Stdout, logLevel, "Global variables use {0} bytes of dynamic memory.", strconv.Itoa(dataSize))
+		}
+	}
+
+	if textSize > maxTextSize {
+		return errors.New("Sketch too big; see http://www.arduino.cc/en/Guide/Troubleshooting#size for tips on reducing it.")
+	}
+
+	if maxDataSize > 0 && dataSize > maxDataSize {
+		return errors.New("Not enough memory; see http://www.arduino.cc/en/Guide/Troubleshooting#size for tips on reducing your footprint.")
+	}
+
+	if properties["build.warn_data_percentage"] != "" {
+		warnDataPercentage, _ := strconv.Atoi(properties["build.warn_data_percentage"])
+		if maxDataSize > 0 && dataSize > maxDataSize*warnDataPercentage/100 {
+			logger.Fprintln(os.Stdout, logLevel, "Low memory available, stability problems may occur.")
+		}
+	}
+
+	return nil
+}
+
+func execSizeReceipe(properties properties.Map, logger i18n.Logger, textSize *int, dataSize *int, eepromSize *int) error {
+
+	out, err := builder_utils.ExecRecipe(properties, constants.RECIPE_SIZE_PATTERN, false, false, false, logger)
+
+	if err != nil {
+		return errors.New("Couldn't determine program size: {0}")
+	}
+
+	// force multiline match prepending "(?m)" to the actual regexp
+
+	if len(properties["recipe.size.regex"]) > 0 {
+		textRegexp := regexp.MustCompile("(?m)" + properties["recipe.size.regex"])
```

Comment:
`regexp.Compile` should be used instead of `regexp.MustCompile` because is not said that `recipe.size.regex` contains a valid regular expression. In case of error we should print a warning and ignore the regexp as it was missing.